Feature/m0078 ddg 308 tutorial debug applications (#334)
MicroEJ/docs: Tutorials/tutorialDiscoverEmbeddedDebuggingToolsAndMethods.rst

.. _tutorial_discover_embedded_debugging_tools_and_methods.tools.platform_qualification:

Platform Qualification
~~~~~~~~~~~~~~~~~~~~~~

The Platform Qualification Tools (PQT) project provides the tools required to validate each component of a MicroEJ Platform.
After porting or adding a feature to a MicroEJ Platform, it is necessary to validate its integration.

The project is available on GitHub: https://github.com/MicroEJ/PlatformQualificationTools

.. figure:: ../PlatformDeveloperGuide/images/overview-platform-qualification.png
   :align: center
   :scale: 80%

   Platform Qualification Overwiew

Please refer to the :ref:`platform_qualification` documentation for more information.

.. _tutorial_discover_embedded_debugging_tools_and_methods.use_case_1:

Use Case 1:  Debugging an UI Application Freeze
-----------------------------------------------

When an application User Interface freezes and becomes unresponsive, in most cases, one of the following conditions applies:

- An unrecoverable system failure occurred, like a HardFault, and the RTOS tasks are not scheduled anymore.

> - The RTOS task that runs the MicroEJ runtime is never given CPU time (suspended or blocked).

Reviewer: the explanation of green thread appears after (as a reminder...) 
I think this should be explained before.

Author: I disagree.
First, because in the introduction I want the user to quickly understand if it is applicable to their use case.
Second, because the green thread explanation is not important there, it's a technical detail but all developers should be able to understand the sentence "RTOS task that runs the MicroEJ runtime".  The green thread explanation helps to understand why it may block the application.  But from an application or firmware developer point of view, both can understand that if the runtime has no CPU time, the application won't be executed.

> - The RTOS task that runs the MicroEJ runtime is never given CPU time (suspended or blocked).

Reviewer: review usage of "MicroEJ runtime" everywhere: replace by "Core Engine" when speaking about the RTOS "task"
```suggestion
- The RTOS task that runs the Core Engine is never given CPU time (suspended or blocked).
```

Author: Is MicroEJ Core Engine ok here?  To make sure the distinction between the RTOS and the Core Engine is clear.

Author: It replaced `MicroEJ runtime`  with `MicroEJ Core Engine` here (and here only, I let `Core Engine` at the other location)

- The RTOS task that runs the MicroEJ runtime is executing never-ending native code (infinite loop in native implementation for example).
- A Java method executes a long-running operation in the MicroUI thread (also called Display Pump thread).
- The application code is unable to receive or process user input events.

The following sections explain how to instrument the code to locate the issue when the UI freeze occurs.
The steps followed are:

> 1. Check if the RTOS Scheduler properly schedules the MicroEJ runtime task.
> 2. Check if the Java Scheduler properly schedules all Java threads.
> 3. Check if the Java Scheduler properly schedules the MicroUI thread.

Reviewer: ```suggestion
1. Check if the RTOS properly schedules the MicroEJ runtime task.
2. Check if the Core Engine properly schedules all Java threads.
3. Check if the Core Engine properly schedules the MicroUI thread.
```

4. Check if Input Events are properly processed.

.. note::
   - The checks of the schedulers are possible with :ref:`systemview` and :ref:`MicroUI Debug Traces <microui_traces>`.
   - The Input Events check is possible with the :ref:`LLUI_INPUT_dump API <section_inputs_eventbuffer>`.

> Check RTOS Scheduler Liveness

Reviewer: ```suggestion
Check RTOS Tasks Scheduling
```

on standard output, blink a LED, use SystemView, etc.).

If the heartbeat is still running when the UI freeze occurs, we can
go a step further and check whether the MicroEJ runtime is still
scheduling Java threads or not.

If you use task priorities for the RTOS tasks management, ensure that
the priority of the RTOS task is equal or lower than the priority of
the MicroEJ runtime task.

If the RTOS task of the heartbeat doesn't run when:

- the priority is the highest than any other tasks, then the RTOS
  scheduler is not scheduling anything.
- the priority is the same as the MicroEJ runtime and other tasks with
  a higher priority exists, then one or more RTOS tasks are causing
  starvation by taking all the resources.

..
   @startuml
   if (heartbeat task runs\nwith highest priority) then (no)
     #pink:RTOS scheduler not working;
     kill
   else (yes)
     if (Heartbeat task runs\nwith same priority\nas MicroEJ Runtime) then (no)
       #pink:MicroEJ Runtime is starving;
       kill
     else (yes)
       :MicroEJ Runtime is running;
       kill
     endif
   endif
   @enduml  

.. image:: images/tuto_microej_debug_ui_freeze_rtos_task_heart_beat_priority.png

> Check Java Scheduler Liveness

Reviewer: ```suggestion
Check Java Threads Scheduling
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> As a reminder, the architecture of the MicroEJ runtime is called green thread architecture, it defines a multi-threaded environment without relying on any native RTOS capabilities. 
> Therefore, the whole Java world runs in one single RTOS task.
> Read more about the MicroEJ Runtime architecture in the :ref:`Platform Developer Guide <core_engine>`.

Reviewer: Remove "architecture" word to describe the threading model
```suggestion
As a reminder, the threading model implemented by MicroEJ runtime is called green thread: it defines a multi-threaded environment without relying on any native RTOS capabilities. 
Therefore, the whole Java threads run in a single RTOS task.
For more details, please refer to the :ref:`MicroEJ Core Engine<core_engine>` section.
```

* A native (C) function has a bug and write to an incorrect memory location
* A native stackoverflow
* A native heap overflow
* A device incorrectly initialized or misconfigured.
* ...

When the hardfault occurs in the MicroJVM task, the VM task heap or stack may be corrupted.
Add ``LLMJVM_checkIntegrity`` call in checkpoints of the BSP code to identify the timeslot of the memory corruption.
Typically, you can check a native with:

.. code-block:: java

   void Java_com_mycompany_MyClass_myNativeFunction(void) {
   	int32_t crcBefore = LLMJVM_checkIntegrity();
   	myNativeFunctionDo();
   	int32_t crcAfter = LLMJVM_checkIntegrity();
   	if(crcBefore != crcAfter){
   		// Corrupted memory in MicroJVM virtual machine internal structures
   		while(1);
   	}
   }

Investigation
~~~~~~~~~~~~~

Determine which memory regions are affected and determine which components are responsible for the corruption.

* List all memory available and their specifics

    * Access mode (addressable, DMA, ...)
    * Cache mechanism? L1, L2

* Is low-power enabled for CPU and peripherals? Is the memory disabled/changed to save power?
* Get the memory layout of the project

>     * What are the code sections for BSP and Java

Reviewer: Java what ? (Java must be followed by a word)

Author: Replaced by Application

    * Where are the BSP stack and heap? What about the Application stack and heap?
    * Where are the Java immortals heap?
    * Where are the Java strings?
    * Where is the MicroEJ UI buffer?

>     * Besides the Java immortals, what are the other intersection point between the Java and the BSP? (e.g. a temporary RAM buffer for JPEG decoder).

Reviewer: same comment Java what ?

    * Please refer to the :ref:`Core Engine Link section <core_engine_link>` to locate the Application sections and to the :ref:`application_options` for their sizes.

> * Implement a CRC of the *hot sections* when entering/leaving all natives. *Hot Sections* are memory sections used by both the Java and native code (e.g. C or ASM).

Reviewer: ```suggestion
* Implement a CRC of the *hot sections* when entering/leaving all natives. *Hot Sections* are memory sections used by both Java code and native code (e.g. C or ASM).
```

* Move the C stack at the beginning of the memory to trigger a crash when it overflows (instead of corrupting the memory).

When a Hardfault Occurs
~~~~~~~~~~~~~~~~~~~~~~~

Extract Information and Coredump
++++++++++++++++++++++++++++++++

Attach an embedded debugger and get the following information:

* stack traces and registers information for each stack frame
* memory information

    * the whole memory if possible
    * otherwise, get the *hot sections* 

        * BSP and Java heap and stack
        * UI buffer
        * immortals heap

>         * sections where the Java and BSP are working together

Reviewer: Java what ? (Java must be followed by a word)

Author: Added "Java application"
